### PR TITLE
Schedule Enhancements

### LINKED ISSUE #473: [Scheduler] Add support for process time accounting
# Motivation

In most Unix systems, the process schedulers are responsible for managing how long a process can run before it is potentially preempted in favor of another runnable process. This mechanism involves decrementing a process's timeslice on each tick of the system clock.

However, in the current implementation of Jinux's scheduler, there is a notable deficiency – it lacks support for process time accounting. This absence of time accounting prevents the scheduler from effectively implementing process preemption for normal processes.

Once we have added the necessary process time accounting to Jinux's scheduler, we can  integrate more effective scheduling algorithms to further improve the system's performance and responsiveness.

# Design Draft
` !todo()`

### LINKED ISSUE #478: [Preemption] A Destined Panic in `preempt()`?
# Problem?

After `disable_irq`(in line 68), a `DisabledLocalIrqGuard` is held, which contains a `DisablePreemptGuard` that has increased `PREEMPT_COUNT`.

https://github.com/jinzhao-dev/jinux/blob/0e7f279bd6b1ee4fe564c999d953a5dc1e520bf3/framework/jinux-frame/src/task/processor.rs#L66-L81

https://github.com/jinzhao-dev/jinux/blob/0e7f279bd6b1ee4fe564c999d953a5dc1e520bf3/framework/jinux-frame/src/task/processor.rs#L165-L169

The `disable_irq` is still held when entering `switch_to_task(next_task)` in line 80, and the preemptible check (in line 91) may not succeed?

https://github.com/jinzhao-dev/jinux/blob/0e7f279bd6b1ee4fe564c999d953a5dc1e520bf3/framework/jinux-frame/src/task/processor.rs#L90-L98

https://github.com/jinzhao-dev/jinux/blob/0e7f279bd6b1ee4fe564c999d953a5dc1e520bf3/framework/jinux-frame/src/task/processor.rs#L148-L150

# Solution

If the problem indeed exists, two rudimentary solutions may be:

- drop the `disable_irq` before `switch_to_task(next_task)`;
- [move the preemptible check out from `switch_to_task`](https://github.com/fgh1999/jinux/blob/cf902e58def380b993039d85016148ad8577200d/framework/jinux-frame/src/task/processor.rs#L99-L103), and add [the check](https://github.com/fgh1999/jinux/blob/cf902e58def380b993039d85016148ad8577200d/framework/jinux-frame/src/task/processor.rs#L55-L73) manually before the switch.

### PR DESCRIPTION
**Pull request by fgh1999**
# Preemption Check Fix

An `active` flag is added into `PreemptInfo` to isolate the increasement of `PREEMPT_COUNT` caused by:

- locks and irqs,
- or, preemption disalbing during the scheduling (previously, in `preempt`).

Fix #478.

Preemptible checks are refactored accordingly.

# Tick & Timeslice

The `scheduler_tick` handler is added to the system to enable clock irq-driven preemptive scheduling.

tick

On each `TICK` update, the `tick` method implemented by the the scheduler is triggered to decrease the timeslice of the current task(from `PROCESSOR`). The `tick` methods have nothing to do with the current scheduler but are required by the O(1) and CFS schedulers.

Fix #473.

# Refactor `schedule`

`preempt` in the user loop is deprecated. Instead, `schedule` is refactored to take the current task (held by the global `PROCESSOR`) into the consideration of picking the *next* task, which in fact invloves preemption.

The global PROCESSOR are now guarded by SpinLock.
Some unnecessary irq-disable operations are downgraded to reduce performance cost.

Some new hooks are added into the `Scheduler` trait. The current scheduler, `PreemptScheduler`, is refactored accordingly into `sched::scheduler::vanilla::PreemptiveRRScheduler`.

# ~Child Runs First~

`CHILD_RUN_FIRST`, a config option, is added into the aster-frame to bypass the implicit order dependency we mentioned in #518, along with two tmp fixes:

- let child tasks run first explicitly in `Task::run`;
- #532 for #530 and #531.

Discussions about *if a freshly forked child should run before the parent continues execution* are also taken in Linux v2. The behavior is controlled by an argument named `sched_child_runs_first` in Linux.

- The argument was set to 1 for several minor versions in v2. Supporters believed that the child-run-first policy could reduce ticks wasted on unnecessary copy-on-write page copies and give "higher priority" to newly forked processes as often the case.
- However, the argument is defaulted as 0 in v2.6.32 and has remained unchanged since then. Linus made the [decision](https://yarchive.net/comp/linux/child-runs-first.html) based on considerations regarding TLB, cache, and prior bash bugs.

The `CHILD_RUN_FIRST` in aster-frame is currently defaulted to `true` because we have not been able to alleviate the extremely severe performance reduction. #563 and #564 may help.

Update: the option won't be provided to control the behavior in the first exec order of a new child task.

### CODE REVIEW

File: framework/aster-frame/src/task/scheduler.rs [hunk 1]
```diff
@@ -1,62 +1,146 @@
...
+    /// according to the time update.
+    ///
+    /// # Arguments
+    ///
+    /// * `cur_task` - the task to be charged, must be held by the processor, and not in runqueue
+    ///
+    /// # Returns
+    ///
+    /// `true` if the `task` need a rescudule.
+    fn tick(&self, cur_task: &Arc<T>) -> bool;
```
Comment (jellllly420) [inline framework/aster-frame/src/task/scheduler.rs:49]: I don't get why we need the return value and where we are using it currently.

Comment (fgh1999) [reply framework/aster-frame/src/task/scheduler.rs:49]: Yes, it's of no use anymore and will be discarded.

File: framework/aster-frame/src/lib.rs [hunk 2]
```diff
@@ -62,6 +62,8 @@ pub fn init() {
     boot::init();
     vm::init();
     trap::init();
+    task::init_processor();
+    task::init_scheduler();
```
[the remark this answers is not in the document]
Comment (fgh1999) [reply framework/aster-frame/src/lib.rs:66]: ok

File: framework/aster-frame/src/task/processor.rs [hunk 3]
```diff
@@ -36,152 +33,150 @@ impl Processor {
...
 
 pub fn current_task() -> Option<Arc<Task>> {
-    PROCESSOR.lock().current()
+    PROCESSOR.get().unwrap().lock_irq_disabled().current()
 }
 
-pub(crate) fn get_idle_task_cx_ptr() -> *mut TaskContext {
-    PROCESSOR.lock().get_idle_task_cx_ptr()
+#[inline]
+fn current_task_without_irq_disabling() -> Option<Arc<Task>> {
```
[the remark this answers is not in the document]
Comment (fgh1999) [reply framework/aster-frame/src/task/processor.rs:55]: ok

File: framework/aster-frame/src/task/processor.rs [hunk 4]
```diff
@@ -36,152 +33,150 @@ impl Processor {
...
-    static ref PROCESSOR: Mutex<Processor> = Mutex::new(Processor::new());
+static PROCESSOR: spin::Once<SpinLock<Processor>> = spin::Once::new();
+
+pub fn init() {
+    PROCESSOR.call_once(|| SpinLock::new(Processor::new()));
 }
 
 pub fn take_current_task() -> Option<Arc<Task>> {
-    PROCESSOR.lock().take_current()
+    PROCESSOR.get().unwrap().lock_irq_disabled().take_current()
```
[the remark this answers is not in the document]
Comment (fgh1999) [reply framework/aster-frame/src/task/processor.rs:47]: 1. `.get_unchecked()` needs `unsafe`. I've no idea whether it is acceptable.
2. `take_current_task()` is not used and I don't know what it is intended for. It is upgraded to `lock_irq_disabled` for the following reasons:
  - It is exported as a **public** function of the `PROCESSOR`.
  - For now, the `PROCESSOR` is only used in scheduling. All the operations during scheduling are performed in an irq-disabled context.

File: framework/aster-frame/src/task/processor.rs [hunk 5]
```diff
@@ -36,152 +33,150 @@ impl Processor {
...
-    if !scheduler.should_preempt(&curr_task) {
-        return;
-    }
-    let Some(next_task) = scheduler.dequeue() else {
-        return;
-    };
-    drop(scheduler);
-    switch_to_task(next_task);
+/// Irq disabled version of `get_idle_task_ctx_ptr`.
+fn idle_task_ctx_ptr() -> *mut TaskContext {
```
[the remark this answers is not in the document]
Comment (fgh1999) [reply framework/aster-frame/src/task/processor.rs:67]: The func seems tedious, I'll remove it.

File: framework/aster-frame/src/task/processor.rs [hunk 6]
```diff
@@ -36,152 +33,150 @@ impl Processor {
...
     }
     pub fn set_current_task(&mut self, task: Arc<Task>) {
-        self.current = Some(task.clone());
+        self.current = Some(task);
     }
 }
 
-lazy_static! {
-    static ref PROCESSOR: Mutex<Processor> = Mutex::new(Processor::new());
+static PROCESSOR: spin::Once<SpinLock<Processor>> = spin::Once::new();
```
[the remark this answers is not in the document]
Comment (fgh1999) [reply framework/aster-frame/src/task/processor.rs:40]: The `PROCESSOR` is accessed during scheduling, e.g., in `schedule()`. `Mutex` may `yield_now()` if it failed to acquire the lock at once.
It is kind of dangerous to invoke `schedule()`(from `yield_now()`) inside a `schedule()`.

BTW, we may also need to replace the mutex in `Task` (`task_inner: Mutex<TaskInner>`) with spinlock.

File: framework/aster-frame/src/task/preempt.rs [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+    fn is_preemptible(&self) -> bool {
+        // todo: may cause inconsistency
+        !self.in_preemption() && !self.in_atomic()
+    }
+
+    /// Whether the current CPU is in atomic context,
+    /// which means it holds some locks or is in IRQ context.
+    fn in_atomic(&self) -> bool {
+        self.num() != 0
```
[the remark this answers is not in the document]
Comment (fgh1999) [reply framework/aster-frame/src/task/preempt.rs:70]: You are right. thx.

File: framework/aster-frame/src/task/processor.rs [hunk 8]
```diff
@@ -36,152 +34,144 @@ impl Processor {
...
+            &mut cur_task.inner_exclusive_access().ctx as *mut TaskContext
+        }
+    };
+
+    let next_task_ctx = &next_task.inner_ctx() as *const TaskContext;
+    PROCESSOR.lock().set_current_task(next_task);
+    unsafe {
+        context_switch(current_task_ctx, next_task_ctx);
     }
+    activate_preempt();
```
[the remark this answers is not in the document]
Comment (fgh1999) [reply framework/aster-frame/src/task/processor.rs:172]: @jellllly420 You are right. thx.

**Pull request closed without merging**